annotation-processor/src/main/kotlin/pl/touk/krush/source/MappingsGenerator.kt

```diff
@@ -127,37 +209,185 @@ abstract class MappingsGenerator : SourceGenerator {
+                        ClassName("kotlin.collections", "MutableSet")
+                            .parameterizedBy(ANY)
+                    )
+            )
+
+        val func = FunSpec.builder("addSubEntitiesTo${entity.name}")
+            .receiver(ResultRow::class)
+            .addParameter(entityParamName, entityType.toImmutableKmClass().toClassName().copy(nullable = true))
+            .addParameter(
+                ParameterSpec.builder("entityStore", entityStoreType)
+                    .defaultValue("mutableMapOf()")
+                    .build()
+            )
+            .addParameter(
+                ParameterSpec.builder("selfReferenceRequests", selfReferenceRequestType)
+                    .defaultValue("mutableMapOf()")
+                    .build()
+            )
+
+        func.addStatement("if($entityParamName == null) return")
+
+        // Recursively add info to every related O2O entity
+        entity.getAssociations(ONE_TO_ONE).forEach { oneToOneAssoc ->
+            func.addComment("Add sub-elements contained in this row to ${oneToOneAssoc.name}")
+            func.addStatement("addSubEntitiesTo${oneToOneAssoc.target.simpleName}($entityParamName.${oneToOneAssoc.name}, entityStore)")
+        }
+
+        // M2M and M2O relations are represented as lists. When such a list contains multiple entities, those entities
+        // are spread over multiple rows. Here, we figure out whether this row contains a new sub-entity in a list and
+        // add it to that list if necessary. Otherwise, we just recursively search each entity for sub-sub-entities that
+        // might be new in this row, etc.
+        entity.getAssociations(ONE_TO_MANY, MANY_TO_MANY).forEach { setAssoc ->
+
+            if(!setAssoc.isSelfReferential) {
+                val attrValName = "${setAssoc.name.asVariable()}Attr"
+                val newEntityValName = "new${setAssoc.target.simpleName}"
+
+                val targetTypeName = "${setAssoc.target.simpleName}"
+
+                func.apply {
+                    // Allowing a null id here allows users to not include a join with the other table if they don't
+                    // need the relation-lists to be populated
+                    addStatement("val ${setAssoc.name.asVariable()}Id = ${idReadingBlock(setAssoc.targetId, setAssoc.targetTable, nullable = true)}")
+                    addStatement("if(${setAssoc.name.asVariable()}Id != null) {")
+
+
+                    addStatement("\tval $attrValName = $entityParamName.${setAssoc.name.asVariable()} as MutableList<$targetTypeName>")
+                    addStatement("\tval ${attrValName}LastElement = $attrValName.lastOrNull()")
+
+                    addStatement("\tif(${setAssoc.name.asVariable()}Id != ${attrValName}LastElement?.${setAssoc.targetId.name}) {")
+
+                    addComment("\t\tIf the sub-entity is new, create a new object for it")
+                    addStatement("\t\tval $newEntityValName = to$targetTypeName()")
```

Comment:
Why you don't pass `entityStore` here?

Reply:
Good catch! I totally should. Same thing for the `selfReferenceRequests` case.

Reply:
I changed it locally but now it fails with `StackOverflowError' in bi-directional mappings - maybe we need same logic as in bidirectional one-to-one case?

Reply:
Ah, that's the old problem of "doing it's job too well" again - those happen because `toString()` loops endlessly between the objects that now (correctly) point to each other.
I had forgotten that was a thing with the previous real-reference-implementation. It pains me to have to "make it worse", but manually changing `toString()` isn't something we should ask of the user. I agree that we'd have to manipulate the `entityStore` in a similar way to avoid this error - probably something like putting a copy of the @OneToMany entity with an empty list in it just before it gets requested by the @ManyToOne entity.

Reply:
When we're aat this topic, I think `krush` should be more conservative filling data in result model. I'm thinking about using join aliases as a clue what relations should we fill and what leave empty. 
I dig through Micronaut's data jdbc code, and here is how you acheive it there:
https://github.com/micronaut-projects/micronaut-data/blob/fc40a6ce7378d4de67d3407074de51200526be51/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/one2many/MultiOneToManyJoinSpec.groovy#L129

Reply:
What's the advantage of not filling in that data? I don't think it would improve performance by much to read less data, and if a column is included in a query I think it should be included in the object model as well.
I am not familiar with Micronaut, so I can't really tell what the aliases are supposed to accomplish in the linked code.